### ISSUE TITLE
in my sails js application my show and edit function is working without # in id but not working with # id

### ISSUE THREAD

**Issue by gmchaturvedi1** [balderdashy/sails#2876]
i have asked the same question to sails-orientdb team they told this is because the application.Kindly help me how to solve it .

**Comment by gmchaturvedi1** [comment]
mycode is controller:

/**
 * UserController
 *
 * @description :: Server-side logic for managing users
 * @help        :: See http://links.sailsjs.org/docs/controllers
 */

module.exports = {
  'new':function(req,res){
       res.view('user/new');

  },
  create1:function(req,res,next){

    User.create(req.params.all(),function createUser(err,user){
      if(err){
        console.log(err);
        req.session.flash={
          err:err
        }
        return res.redirect('/user/new');
      }
      req.session.authenticated=true;
      req.session.User=user;
      res.redirect('user/show/'+user.id);

    })

  },
  show:function(req,res,next){

    User.findOne(req.param('id'),function(err,user){

      if (err) return next(err);
      if(!user) return next();
      res.view('user/show/',{user:user});
    });
  },
  index:function(req,res,next){

    User.find(function(err,users){
      if(err) return next(err);
      res.view('user/index',{users:users});

    })
  },
  edit:function(req,res,next){
    User.findOne(req.param('id'),function editUser(err,user){
      if(err) return next(err);
      res.view('user/edit',{user:user})

    })

  },
  update:function(req,res,next){
    User.update(req.param('id'),req.params.all(),function updateUser(err){
      if(err){
        return res.redirect('/user/edit/'+req.param('id'));
      }
      res.redirect('/user/show/'+req.param('id'));
    })
  },
  destroy:function(req,res,next){
    User.findOne(req.param('id'),function editUser(err,user){
      if(err) return next(err);
      if(!user) return next("User does not exist");

      User.destroy(req.param('id'),function destroyUser(err){
        if(err) return next(err);
        res.redirect('/user')
      })

    })


  }

};

model:

/**
* User.js
*
* @description :: TODO: You might write a short summary of how this model works and what it represents here.
* @docs        :: http://sailsjs.org/#!documentation/models
*/

module.exports = {

  attributes: {

    name:{
      type:'string'
    },
    title:{
      type:'string'
    },
    email:{
      type:'string',
      required:true

    },
    encryptedPassword:{
      type:'string'
    }

  },
  beforeCreate:function(values,next){
    if(!values.password||values.password!=values.confirmation){
      return next({err:"Password does Not math password confirmation"})
    }
    require('bcrypt').hash(values.password,10,function passwordEncrypted(err,encryptedPassword){
      if(err) return next(err);
      values.encryptedPassword=encryptedPassword;
      next();
    })

  }
};

and route:
/**
 * Route Mappings
 * (sails.config.routes)
 *
 * Your routes map URLs to views and controllers.
 *
 * If Sails receives a URL that doesn't match any of the routes below,
 * it will check for matching files (images, scripts, stylesheets, etc.)
 * in your assets directory.  e.g. `http://localhost:1337/images/foo.jpg`
 * might match an image file: `/assets/images/foo.jpg`
 *
 * Finally, if those don't match either, the default 404 handler is triggered.
 * See `api/responses/notFound.js` to adjust your app's 404 logic.
 *
 * Note: Sails doesn't ACTUALLY serve stuff from `assets`-- the default Gruntfile in Sails copies
 * flat files from `assets` to `.tmp/public`.  This allows you to do things like compile LESS or
 * CoffeeScript for the front-end.
 *
 * For more information on configuring custom routes, check out:
 * http://sailsjs.org/#/documentation/concepts/Routes/RouteTargetSyntax.html
 */

module.exports.routes = {

  /***************************************************************************
  *                                                                          *
  * Make the view located at `views/homepage.ejs` (or `views/homepage.jade`, *
  * etc. depending on your default view engine) your home page.              *
  *                                                                          *
  * (Alternatively, remove this and add an `index.html` file in your         *
  * `assets` directory)                                                      *
  *                                                                          *
  ***************************************************************************/

  '/': { view: 'homepage'},
  'get /user/new':'UserController.new',
  'post /user/create1':'UserController.create1',
  'get /user/show/:id':'UserController.show',
  'get /user/edit/:id':'UserController.edit',
  'get /session/new':'SessionController.new',
  'post /session/create1':'SessionController.create1'

  /***************************************************************************
  *                                                                          *
  * Custom routes here...                                                    *
  *                                                                          *
  *  If a request to a URL doesn't match any of the custom routes above, it  *
  * is matched against Sails route blueprints. See `config/blueprints.js`    *
  * for configuration options and examples.                                  *
  *                                                                          *
  ***************************************************************************/

};

**Comment by tjwebb** [comment]
> is working without # in id but not working with # id

It's difficult to understand what you're asking because you apparently don't have the time to write in complete sentences. Are you saying that when your `id` contains the `#` character, you receive some error?

`not working` is not an error message. Post *exactly* what is happening.

**Comment by gmchaturvedi1** [comment]
when id is containg # then error code is showing and when i am removing that its works well i have asked the same thing from sails-orientdb people they told me that i need to adjust my code to work well

**Comment by gmchaturvedi1** [comment]
in OrientDB @rid is like #12:01

**Comment by tjwebb** [comment]
> when id is containg # then error code is showing

Again. Post the error message.

**Comment by gmchaturvedi1** [comment]
console is showing :verbose: Sending 404 ("Not Found") response

**Comment by tjwebb** [comment]
Sorry, I just don't know how to help you. You're not posting any errors. "404" is not an error, it just means you're requesting something that doesn't exist.

By default, the `id` field of a model is an integer. Inserting a `#` into your `id` doesn't make a lot of sense in the first place, and I would just use integers.

**Comment by gmchaturvedi1** [comment]
i am not adding this is by default orient db created  and map to id .sorry to bother you

**Comment by tjwebb** [comment]
> i am not adding this is by default orient db created and map to id

Why did the orientdb project say that this was not an orientdb issue?

**Comment by dmarcelino** [comment]
appscot/sails-orientdb#87

Tl;dr: `req.param('id')` is `undefined`.

**Comment by gmchaturvedi1** [comment]
Dear All,

i found the problem that findone method is not working with the $id but when i am using only find method it is working.

**Comment by dmarcelino** [comment]
@gmchaturvedi1, I'm afraid you are mistaken. Considering what you've told us in appscot/sails-orientdb#87, your id is `undefined` which is not a valid argument for `findOne()`.

sails-orientdb does support `findOne(id)`, there is an [automated test for this scenario](https://github.com/balderdashy/waterline-adapter-tests/blob/master/interfaces/semantic/findOne.test.js#L52):
``` javascript
    it('should work with just an id passed in', function(done) {
      Semantic.User.findOne(id, function(err, user) {
        assert(!err);
        assert(user.first_name === 'findOne test');
        done();
      });
    });
````
and sails-orientdb passes it.

The reason why `find()` is working for you is because `find(undefined)` returns all records for a given collection.

You can read more about the differences between `find()` and `findOne()` on the [Waterline Documentation](https://github.com/balderdashy/waterline-docs/blob/159a7ad89df0d47ff3c16e61456766245889ffbb/query-methods.md#find-criteria-callback-)

**Comment by mikermcneil** [comment]
@dmarcelino @tjwebb excellent answers, thanks guys.
